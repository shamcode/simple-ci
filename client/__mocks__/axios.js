const METHODS = [
    'get',
    'post',
    'put',
    'delete'
];

const FIXTURES = {};
const MOCKS = {};

function mockMethodFactory( method ) {
    return jest.fn().mockImplementation(
        ( url ) => {
            if ( !FIXTURES[ method ].hasOwnProperty( url ) ) {
                throw new Error( `Missing fixture for ${method.toUpperCase()} url = "${url}"` );
            }
            const { success, data } = FIXTURES[ method ][ url ];
            return Promise[ success ? 'resolve' : 'reject' ]( {
                data
            } );
        }
    );
}

METHODS.forEach(
    method => {
        MOCKS[ method ] = mockMethodFactory( method );
        FIXTURES[ method ] = {};
    }
);

const defaultMocksData = ( () => {
    const data = {};
    data.chain = {
        id: 2,
        name: 'Test chain',
        command: 'echo 1'
    };
    data.project = {
        id: 1,
        name: 'Test',
        cwd: '/tmp/',
        chains: [ data.chain ]
    };
    return data;
} )();

export default {
    defaultMocksData: defaultMocksData,
    mocks: MOCKS,
    create: jest.fn().mockImplementation( () => {
        return {
            ...MOCKS,
            interceptors: {
                request: {
                    use: () => {}
                },
                response: {
                    use: () => {}
                }
            }
        };
    } ),
    use( method, url, data, success = true ) {
        if ( !METHODS.includes( method ) ) {
            throw new Error( `Unknown method: "${method}"`);
        }
        FIXTURES[ method ][ url ] = {
            data,
            success
        };
        return this;
    },
    useDefaultMocks() {
        this
            .use( 'get', '/projects', [ defaultMocksData.project ] )
            .use( 'get', '/projects/1', defaultMocksData.project )
            .use( 'get',  '/chains/2', defaultMocksData.chain );
        return this;
    }
};