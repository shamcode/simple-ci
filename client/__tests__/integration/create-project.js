import pretty from 'pretty';
import axios from 'axios';
import ShamUI, { DI } from 'sham-ui';
import controller from '../../src/controllers/main';
jest.mock( 'axios' );

const flushPromises = () => new Promise( resolve => setImmediate( resolve ) );

beforeEach( () => {
    jest.resetModules();
    jest.clearAllMocks();
} );

it( 'create project', async() => {
    expect.assertions( 7 );

    const postMock = jest.fn().mockReturnValue( Promise.resolve() );
    axios.create.mockImplementation( () => {
        return {
            get: jest.fn().mockReturnValue( Promise.resolve( {
                data: []
            } ) ),
            post: postMock,
            interceptors: {
                response: {
                    use: () => {}
                }
            }
        };
    } );

    DI.bind( 'widget-binder', controller );
    const UI = new ShamUI();
    UI.render.FORCE_ALL();
    await flushPromises();

    const body = document.querySelector( 'body' );
    body.querySelector( '.project-card-create' ).click();
    expect( pretty( body.innerHTML ) ).toMatchSnapshot();

    const formData = {
        name: 'Test name',
        cwd: 'test cwd'
    };
    body.querySelector( '[name="name"]' ).value = formData.name;
    body.querySelector( '[name="cwd"]' ).value = formData.cwd;
    body.querySelector( '[type="submit"]' ).click();
    expect( postMock ).toHaveBeenCalledTimes( 1 );
    expect( postMock.mock.calls[ 0 ][ 0 ] ).toBe( '/projects' );
    expect( Object.keys( postMock.mock.calls[ 0 ][ 1 ] ) ).toEqual( [ 'name', 'cwd' ] );
    expect( postMock.mock.calls[ 0 ][ 1 ].name ).toEqual( formData.name );
    expect( postMock.mock.calls[ 0 ][ 1 ].cwd ).toEqual( formData.cwd );

    await flushPromises();

    expect( pretty( body.innerHTML ) ).toMatchSnapshot();
} );

