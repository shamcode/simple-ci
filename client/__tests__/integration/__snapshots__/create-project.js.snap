// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`create project 1`] = `
"<header>
  <a class=\\"logo\\" href=\\"\\"> Simple CI </a>
</header>
<!--Header-->
<main>
  <ul class=\\"breadcrumb\\">
    <li>
      <a href=\\"\\"> Projects </a>
    </li>
    <!--if-->
    <li>Create project</li>
  </ul>
  <!--BreadCrumb-->
  <!--breadcrumbs-->
  <!--if-->
  <!--ErrorsPanel-->
  <div class=\\"project-create-form\\">
    <form>
      <div class=\\"group\\">
        <input name=\\"name\\" type=\\"text\\" required=\\"\\"><span class=\\"highlight\\"></span><span class=\\"bar\\"></span>
        <label>Name</label>
      </div>
      <div class=\\"group\\">
        <input name=\\"cwd\\" type=\\"text\\" required=\\"\\"><span class=\\"highlight\\"></span><span class=\\"bar\\"></span>
        <label>Current Work Directory</label>
      </div>
      <div class=\\"group\\">
        <button type=\\"submit\\"><span>Create</span></button>
      </div>
    </form>
  </div>
  <!--ProjectForm-->
  <!--content-->
  <!--if-->
  <!--PageContent-->
</main>
<footer> Sham 2018 </footer>
<!--Footer-->"
`;

exports[`create project 2`] = `
"<header>
  <a class=\\"logo\\" href=\\"\\"> Simple CI </a>
</header>
<!--Header-->
<main>
  <!--breadcrumbs-->
  <!--if-->
  <!--ErrorsPanel-->
  <div class=\\"project-list\\">
    <a class=\\"project-card\\" href=\\"projects/1\\"><span class=\\"name\\">Test</span></a>
    <!--for-->
    <a class=\\"project-card project-card-create\\" href=\\"projects/create\\">
      <div class=\\"create-icon\\"></div>
    </a>
  </div>
  <!--ProjectsList-->
  <!--content-->
  <!--if-->
  <!--PageContent-->
</main>
<footer> Sham 2018 </footer>
<!--Footer-->"
`;
